src/mlpack/core/cereal/pointer_variant_wrapper.hpp

```diff
@@ -0,0 +1,88 @@
+/**
+ * @file core/cereal/pointer_variant_wrapper.hpp
+ * @author Omar Shrit
+ *
+ * Implementation of a boost::variant wrapper to enable the serialization of
+ * the pointers inside boost variant in cereal
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_CORE_CEREAL_POINTER_VARIANT_WRAPPER_HPP
+#define MLPACK_CORE_CEREAL_POINTER_VARIANT_WRAPPER_HPP
+
+#include <cereal/archives/json.hpp>
+#include <cereal/archives/portable_binary.hpp>
+#include <cereal/archives/xml.hpp>
+#include <cereal/types/boost_variant.hpp>
+
+#include <boost/variant.hpp>
+
+#include "pointer_wrapper.hpp"
+
+namespace cereal {
+
+struct save_vistor : public boost::static_visitor<>
+{
+  template<class Archive, class T>
+  void operator()(Archive& ar, T t) const 
+  {
+    ar & CEREAL_POINTER(t);
+  }
+};
+
+struct load_visitor : public boost::static_visitor<>
+{
+  template<class Archive, class T>
+  void operator()(Archive& ar, T t) const 
+  {
+    ar & CEREAL_POINTER(t);
+  }
+};
+
+template<typename VariantType1, typename... VariantTypes>
+class pointer_variant_wrapper
+{
+/*
+ * The objective of this class is to create a wrapper for
+ * boost::variant. 
+ * Cereal supports the serialization of boost::vairnat, but 
+ * we need to serialize it if it holds a raw pointers.
+ */
+ public:
+   pointer_variant_wrapper(boost::variant<VariantType1*, VariantTypes*...>& PointerVar)
+    : PointerVariant(PointerVar)
+  {}
+
+  template<class Archive>
+  void save(Archive& ar) const
+  { 
+    save_vistor s(ar);
+    boost::apply_visitor(s, PointerVariant);
+  }
+
+  template<class Archive>
+  void load(Archive& ar)
+  { 
+    load_visitor l(ar);
+    boost::apply_visitor(l, PointerVariant);
```

Comment:
Are you sure this will work in general?  I think that we need to serialize the type of `PointerVariant`, so that we can load the right type.  You might consider taking a look at how Cereal implements its `boost::variant` support and then imitating that.  After all this should be basically the same, you just need to call `POINTER_WRAPPER()` when you serialize whatever's held since it's a bare pointer.

Reply:
https://github.com/USCiLab/cereal/blob/master/include/cereal/types/variant.hpp

What you've got is close-ish to what they're doing there; I think that a main piece missing is that you have to serialize "index":

https://github.com/USCiLab/cereal/blob/master/include/cereal/types/variant.hpp#L96

The `load_variant` templated function might take a while to understand, but it's a clever way of calling the load visitor with the right template type.  If you like, I can try and walk you through it, just let me know. :+1:

Reply:
Hmm, So basically, I will have to imitate the `boost::variant` implementation in cereal, I have tried to do that before but I did not understand everything there. I will spend more time on their implementation to understand what is happening.

Reply:
@rcurtin I have thought about the `index` since we are serializing pointers. However,  it is only in `std::variant` there is nothing similar in `boost::variant`

Reply:
Sorry, perhaps I linked to the wrong file. https://github.com/USCiLab/cereal/blob/master/include/cereal/types/boost_variant.hpp